analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs

```diff
@@ -18,10 +18,131 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Collections;
+using System.Collections.ObjectModel;
+using SonarAnalyzer.SymbolicExecution.Sonar.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
 
 public abstract class EmptyCollectionsShouldNotBeEnumeratedBase : SymbolicRuleCheck
 {
     protected const string DiagnosticId = "S4158";
     protected const string MessageFormat = "Remove this call, the collection is known to be empty here.";
+
+    private static readonly ImmutableArray<KnownType> TrackedCollectionTypes = ImmutableArray.Create(
+        KnownType.System_Collections_Generic_List_T,
+        KnownType.System_Collections_Generic_HashSet_T,
+        KnownType.System_Collections_Generic_Queue_T,
+        KnownType.System_Collections_Generic_Stack_T,
+        KnownType.System_Collections_ObjectModel_ObservableCollection_T,
+        KnownType.System_Array,
+        KnownType.System_Collections_Generic_Dictionary_TKey_TValue);
+
+    private static readonly ImmutableArray<string> RaisingMethods = ImmutableArray.Create(
```

Comment:
Why are all these immutable arrays instead of normal arrays?

Reply:
Semantics (and maybe performance). I can use normal arrays for simplicity if that's more appropriate.

Reply:
I don't think they are faster by intuition (that's for sure for dictionaries, that are not O(n) but O(log n) as they are AVL trees internally).

Please use just arrays. Or the HashSet (needs measurement)